### PR TITLE
Allow different values than only -1, 0 and 1 to be returned by <=>

### LINKED ISSUE #7276: Allow <=> to not only return -1, 0 or 1
Currently the documentation of [`Comparable`](https://crystal-lang.org/api/0.27.0/Comparable.html) and [`PartialComparable`](https://crystal-lang.org/api/0.27.0/PartialComparable.html) says that classes which include one of them must implement `<=>` and it must return either `1`, `0` or `-1` (or `nil` in case of `PartialComparable`). But why? Is there a specific reason for that? There's not really a benefit behind being so strict. It should probably be changed so that the spaceship is allowed to return either a positive number, zero or a negative number. A good example is `Char`'s [`<=>`](https://crystal-lang.org/api/0.27.0/Char.html#%28other%3AChar%29-instance-method) which already does that.

See also the discussion in: #7269

### PR DESCRIPTION
**Pull request by r00ster91**
Closes #7276 

This changes the spaceship's documentation to allow more values than only `-1`, `0` and `1` to be returned.

### CODE REVIEW

File: src/array.cr [hunk 1]
```diff
@@ -167,11 +167,11 @@ class Array(T)
...
   end
 
-  # Combined comparison operator. Returns *0* if `self` equals *other*, *1* if
-  # `self` is greater than *other* and *-1* if `self` is smaller than *other*.
+  # Combined comparison operator. Returns `0` if `self` equals *other*, `1` if
+  # `self` is greater than *other* and `-1` if `self` is smaller than *other*.
   #
   # It compares the elements of both arrays in the same position using the
-  # `<=>` operator.  As soon as one of such comparisons returns a non-zero
+  # `<=>` operator. As soon as one of such comparisons return a non-zero
```
Comment (RX14) [inline src/array.cr]: returns was correct before

File: src/char.cr [hunk 2]
```diff
@@ -103,10 +103,12 @@ struct Char
     (ord - other).chr
   end
 
-  # Implements the comparison operator.
+  # The comparison operator.
```
Comment (RX14) [inline src/char.cr]: This should be more verbose like the array docs

File: src/partial_comparable.cr [hunk 3]
```diff
@@ -62,5 +62,16 @@ module PartialComparable(T)
     end
```
Comment (r00ster91) [inline src/partial_comparable.cr]: There's `def compare_with(other : T)` above this and it's only used by this module. Seems to be a helper method so shouldn't it be private?

File: src/number.cr [hunk 4]
```diff
@@ -161,9 +161,10 @@ struct Number
     {(self / number).floor, self % number}
   end
 
-  # Implements the comparison operator.
+  # The comparison operator.
   #
-  # See also: `Object#<=>`.
```
Comment (r00ster91) [inline src/number.cr]: Removed because `Object#<=>` no longer exists.

File: src/number.cr [hunk 5]
```diff
@@ -161,9 +161,10 @@ struct Number
...
+  # Returns `-1` if `self` is less than *other*, `0` if `self` equals *other*
```
Comment (straight-shoota) [inline src/number.cr]: IMO readability would be improved if this followed the schema used in the other descriptions to first name the different types of outcome and then the reasons for each.

File: src/partial_comparable.cr (hunk 3 above)
Comment (straight-shoota) [reply src/partial_comparable.cr]: This module is scheduled to be removed in `0.28.0` anyway, so don't bother (see #6611).

Comment (straight-shoota) [reply src/partial_comparable.cr]: This module is scheduled to be removed in 0.28.0 anyway, so don't bother (see #6611).

File: spec/std/array_spec.cr [hunk 6]
```diff
@@ -1180,7 +1180,7 @@ describe "Array" do
     b = [4, 5, 6]
     c = [1, 2]
 
-    (a <=> b).should be < 1
+    (a <=> b).should be < 0
```
[the remark this answers is not in the document]
Comment (r00ster91) [reply spec/std/array_spec.cr]: No, I don't think there was any reason. Looks like a mistake to me. `a` and `b` are never equal.

File: src/comparable.cr [hunk 7]
```diff
@@ -1,22 +1,22 @@
...
   end
 
-  # Compares this object to *other* based on the receiver’s `<=>` method, returning `true` if it returns `-1` or `0`.
+  # Compares this object to *other* based on the receiver’s `<=>` method, returning `true` if it returns a negative number or zero.
   def <=(other : T)
     (self <=> other) <= 0
   end
 
-  # Compares this object to *other* based on the receiver’s `<=>` method, returning `true` if it returns `0`.
+  # Compares this object to *other* based on the receiver’s `<=>` method, returning `true` if it returns zero.
```
Comment (straight-shoota) [inline src/comparable.cr]: I'd prefer to use `0` instead of "zero" everywhere. It's an explicit value and we should use the appropriate literal for that.

File: src/comparable.cr [hunk 8]
```diff
@@ -30,18 +30,20 @@ module Comparable(T)
...
+  # Compares this object to *other* based on the receiver’s `<=>` method, returning `true` if it returns a positive number or zero.
   def >=(other : T)
     (self <=> other) >= 0
   end
 
-  # Comparison operator. Returns `0` if the two objects are equal,
-  # a negative number if this object is considered less than *other*,
+  # The comparison operator.
+  #
+  # Returns zero if the two objects are equal,
```
Comment (straight-shoota) [inline src/comparable.cr]: This should be ordered like all the other cases: negative, zero, positive.

File: src/partial_comparable.cr [hunk 9]
```diff
@@ -62,5 +62,18 @@ module PartialComparable(T)
...
   end
 
+  # The comparison operator.
+  #
+  # Returns a negative number if the object
```
Comment (straight-shoota) [inline src/partial_comparable.cr]: Minor detail, but that's an awkward line break.

File: src/partial_comparable.cr [hunk 10]
```diff
@@ -62,5 +62,18 @@ module PartialComparable(T)
...
+  # is considered to be less than *other*, zero if the two objects are equal,
+  # a positive number if this object is considered to be greater than *other*,
+  # or `nil` if no order can be established.
+  #
+  # Subclasses define this method to provide class-specific ordering.
+  #
+  # ```
+  # # Sort in a descending way
```
Comment (straight-shoota) [inline src/partial_comparable.cr]: This example is unneccesarily hard to follow. Maybe just prepend the same example with ascending order?

File: src/partial_comparable.cr (hunk 9 above)
Comment (r00ster91) [reply src/partial_comparable.cr]: `PartialComparable` is going to be removed anyway.

File: src/comparable.cr (hunk 7 above)
Comment (r00ster91) [reply src/comparable.cr]: I used "zero" because "negative number" and "positive number" is also written in words instead of `-1` and `+1`, so for consistency. But "negative number" and "positive number" is very general while `0` is a unique value so they are still pretty different and `0` reads faster, so I will change that.

**Comment by straight-shoota** [comment]
Thank you @r00ster91
